### DISCUSSION TITLE
Tests setup

### DISCUSSION THREAD

**Discussion by DhivinX** [xeokit/sdk#21] [General]
Hello,

after many hours I managed to run most of the tests. However, I have a problem with tests that use electron in their configuration, mainly with the use of this environment `@kayahr/jest-electron-runner`.

I'm not sure if it's a bug in the `jest-electron-runner` package or if it's a difference in the Jest version, but are there any specific needs to use this package instead of `jsdom`?

@xeolabs I saw that you recently added the `canvas` package, is it used and required somewhere? I'm asking because it makes development much more difficult on Windows, it requires the installation and compilation of the `GTK` package, see: https://www.npmjs.com/package/canvas#compiling

Recently we talked about the report in HTML, currently `test-report.html` is created for each package, is it supposed to stay that way or do you prefer to have tests in one collective place?

**Comment by xeolabs** [comment]
> Hello,
> 
> after many hours I managed to run most of the tests. However, I have a problem with tests that use electron in their configuration, mainly with the use of this environment `@kayahr/jest-electron-runner`.
> 
> I'm not sure if it's a bug in the `jest-electron-runner` package or if it's a difference in the Jest version, but are there any specific needs to use this package instead of `jsdom`?

My intention here was to test @xeokit/viewer with the WebGL-based rendering strategy in @xeokit/webglrenderer. 

Unfortunately the mock HTMLCanvas provided by jsdom doesn't provide a WebGLRenderingContext, otherwise I would have used that. 
 
> @xeolabs I saw that you recently added the `canvas` package, is it used and required somewhere? I'm asking because it makes development much more difficult on Windows, it requires the installation and compilation of the `GTK` package, see: https://www.npmjs.com/package/canvas#compiling

Hmm right. The canvas package was for unit testing the WebGL bits, but If it turns out to be impractical to support the WebGL unit testing I guess we can leave those tests out. 

Or I wonder is it possible to somehow make these dependencies optional, so that they are only installed if the person who checks out the repo actually wants to run the tests, ie. on Linux? I'm guessing that's a bit convoluted, and maybe not possible? 

> Recently we talked about the report in HTML, currently `test-report.html` is created for each package, is it supposed to stay that way or do you prefer to have tests in one collective place?

Yes, at the moment it creates multiple HTML pages, one per tested package, but I was wondering if it were possible to instead compose the results into a single central HTML page, so the results of all tests could be checked at a glance, without navigating.

**Comment by DhivinX** [comment]
Thanks for your replies @xeolabs 

Now I understand everything. I tried to run the tests that require the electron environment, but the package `@kayahr/jest-electron-runner` doesn't work. As you probably already know. 

I will think about how to solve this problem, if I don't come up with anything, we will stick to the `canvas` package. Unfortunately, I don't know if it's possible to exclude a package depending on the OS, but I'll check it out.

I'm already working on the summary report. It may take me a while, if the `jest-html-reporter` configuration does not work, I will create a script that will collect everything in one place - it seems to me a good workaround.

**Reply by xeolabs** [reply]
Cool, thanks very much.

**Comment by DhivinX** [comment]
@xeolabs could you look at https://github.com/capricorn86/happy-dom

**Reply by xeolabs** [reply]
Unfortunately happy-dom doesn't support any WebGL mocking - that's pretty much the core need we have in browser testing.

**Reply by DhivinX** [reply]
@xeolabs Ok, so we have to stick with `canvas`.  The only thing I can do, if possible, is to write a script that will automate the installation of GTK on Windows.  What do you think?

**Comment by xeolabs** [comment]
> @xeolabs Ok, so we have to stick with `canvas`. The only thing I can do, if possible, is to write a script that will automate the installation of GTK on Windows. What do you think?

Yeah that would be good. 

But as a fallback, it's also OK if those particular canvas-dependent tests are something that would only be run on Linux.

**Reply by DhivinX** [reply]
@xeolabs  If I manage to configure eslint, prettier and write a GTK script, what should I do next? Can I help you with anything else?

**Reply by xeolabs** [reply]
Nope I think that's everything I need right now, thanks very much

**Reply by DhivinX** [reply]
Sure, I'll finish these topics. If you need me, feel free to let me know. Thanks!

**Comment by xeolabs** [comment]
Actually, I am getting this issue when running tests, seems Jest can't find a certain mapping, so it fails some tests:

````
npm run test
````

excerpt from the console output:

````
@xeokit/procgen:test:       at Object.<anonymous> (src/buildBoxGeometry.ts:13:15)
@xeokit/procgen:test:       at Object.<anonymous> (src/index.ts:39:14)
@xeokit/procgen:test:       at Object.<anonymous> (tests/buildCylinderGeometry.test.ts:5:16)
@xeokit/procgen:test: 
@xeokit/procgen:test:  FAIL   @xeokit/procgen Tests  tests/buildVectorTextGeometry.test.ts
@xeokit/procgen:test:   ● Test suite failed to run
@xeokit/procgen:test: 
@xeokit/procgen:test:     Configuration error:
@xeokit/procgen:test:     
@xeokit/procgen:test:     Could not locate module @xeokit/core/dist mapped as:
@xeokit/procgen:test:     /home/lindsay/xeolabs/sdk-feb25/packages/core/src/$1.
@xeokit/procgen:test:     
@xeokit/procgen:test:     Please check your configuration for these entries:
@xeokit/procgen:test:     {
@xeokit/procgen:test:       "moduleNameMapper": {
@xeokit/procgen:test:         "/@xeokit\/core(.*)$/": "/home/lindsay/xeolabs/sdk-feb25/packages/core/src/$1"
@xeokit/procgen:test:       },
@xeokit/procgen:test:       "resolver": undefined
@xeokit/procgen:test:     }
@xeokit/procgen:test: 
@xeokit/procgen:test:       11 |  * Creating a {@link @xeokit/scene!SceneMesh} with a box-shaped {@link @xeokit/scene!SceneGeometry}:
@xeokit/procgen:test:       12 |  *
@xeokit/procgen:test:     > 13 |  * ````javascript
@xeokit/procgen:test:          |               ^
@xeokit/procgen:test:       14 |  * TODO
@xeokit/procgen:test:       15 |  * ````
@xeokit/procgen:test:       16 |  *
@xeokit/procgen:test: 
@xeokit/procgen:test:       at createNoMappedModuleFoundError (../../node_modules/.pnpm/jest-resolve@29.7.0/node_modules/jest-resolve/build/resolver.js:759:17)
@xeokit/procgen:test:       at Object.<anonymous> (src/buildBoxGeometry.ts:13:15)
@xeokit/procgen:test:       at Object.<anonymous> (src/index.ts:39:14)
@xeokit/procgen:test:       at Object.<anonymous> (tests/buildVectorTextGeometry.test.ts:5:16)
````

Seems it can't find that @xeokit/core mapping, but it is configured in `jest.config.base.js` though.

**Reply by xeolabs** [reply]
Oops, please disregard my previous message about Jest not finding a mapping - I broke a path wile refactoring a test, then I fixed it.

**Comment by paireks** [comment]
Can we place somewhere in documentation how to run the tests and is there a way to run all of them or only for each module? Locally `npm run test` doesn't work for me, but maybe I don't have something installed? Do I need to install jest separately?

**Reply by DhivinX** [reply]
We have vuepress so we can add documentation in contribution section. But when we using `pnpm` as package manager you should also using `pnpm` instead of `npm` for example `pnpm test` to run tests in repo.

**Reply by DhivinX** [reply]
It is important regards to how `npm` built own dependency tree and `pnpm` has also own way to do this.

**Reply by DhivinX** [reply]
If you want run test only for one package/module please use `turbo run test --filter={module_name}` from root dir. See turborepo docs: https://turbo.build/repo/docs/reference/command-line-reference/run

**Reply by paireks** [reply]
Great thank you! I finally could run them all :) Just added PR with your notes how to run them
